bls/src/keypair.rs

```diff
@@ -0,0 +1,249 @@
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn from_secret(secret: &SecretKey) -> Self {
+        Self(G1Projective::generator() * secret.0)
+    }
+
+    /// Verify a signature against a message and a public key
+    pub fn verify(&self, signature: &SignatureProjective, message: &[u8]) -> bool {
+        Bls::verify(self, signature, message)
+    }
+
+    /// Verify a proof of possession against a public key
+    pub fn verify_proof_of_possession(&self, proof: &ProofOfPossessionProjective) -> bool {
+        Bls::verify_proof_of_possession(self, proof)
+    }
+
+    /// Aggregate a list of public keys into an existing aggregate
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn aggregate_with<'a, I>(&mut self, pubkeys: I)
+    where
+        I: IntoIterator<Item = &'a PubkeyProjective>,
+    {
+        self.0 = pubkeys.into_iter().fold(self.0, |mut acc, pubkey| {
+            acc += &pubkey.0;
+            acc
+        });
+    }
+
+    /// Aggregate a list of public keys
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn aggregate<'a, I>(pubkeys: I) -> Result<PubkeyProjective, BlsError>
+    where
+        I: IntoIterator<Item = &'a PubkeyProjective>,
+    {
+        let mut iter = pubkeys.into_iter();
+        if let Some(acc) = iter.next() {
+            let aggregate_point = iter.fold(acc.0, |mut acc, pubkey| {
+                acc += &pubkey.0;
+                acc
+            });
+            Ok(Self(aggregate_point))
+        } else {
+            Err(BlsError::EmptyAggregation)
+        }
+    }
+}
+
+impl From<PubkeyProjective> for Pubkey {
+    fn from(proof: PubkeyProjective) -> Self {
+        Self(proof.0.to_uncompressed())
+    }
+}
+
+impl TryFrom<Pubkey> for PubkeyProjective {
+    type Error = BlsError;
+
+    fn try_from(proof: Pubkey) -> Result<Self, Self::Error> {
+        let maybe_uncompressed: Option<G1Affine> = G1Affine::from_uncompressed(&proof.0).into();
+        let uncompressed = maybe_uncompressed.ok_or(BlsError::PointConversion)?;
+        Ok(Self(uncompressed.into()))
+    }
+}
+
+impl TryFrom<&Pubkey> for PubkeyProjective {
+    type Error = BlsError;
+
+    fn try_from(proof: &Pubkey) -> Result<Self, Self::Error> {
+        let maybe_uncompressed: Option<G1Affine> = G1Affine::from_uncompressed(&proof.0).into();
+        let uncompressed = maybe_uncompressed.ok_or(BlsError::PointConversion)?;
+        Ok(Self(uncompressed.into()))
+    }
+}
+
+/// A BLS keypair
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Keypair {
+    pub secret: SecretKey,
+    pub public: PubkeyProjective,
+}
+
+impl Keypair {
+    /// Constructs a new, random `Keypair` using a caller-provided RNG
+    pub fn generate<R>(csprng: &mut R) -> Self
+    where
+        R: CryptoRng + RngCore,
+    {
+        let secret = SecretKey::generate(csprng);
+        let public = PubkeyProjective::from_secret(&secret);
+        Self { secret, public }
+    }
```

Comment:
This unfortunately ties us to a certain version of `rand` in the public interface. Makes me realize we should probably deprecate these functions elsewhere in the sdk, so that people generate their own random bytes and pass them into a constructor.

What do you think?

Reply:
Yes good point. I removed the `generate` functions that use the generators directly.